### PR TITLE
Add in-game Ignore + Fix resign

### PR DESCRIPTION
**Pull request by ArchShaman**
- Adds /ignore , /ignorelist and /unignore commands.
- In-game Ignore blocks ALL communication from the player including lobby transmissions (lobby->game).
- Reads from Serverside ignorelist.

### CODE REVIEW

**Comment by GoogleFrog** [comment]
All new widgets should have tabs instead of double spaces.

**Comment by ArchShaman** [comment]
Ok.

**Comment by ArchShaman** [comment]
Finished and noted for future prosperity. Anything else?

File: LuaUI/cawidgets.lua [hunk 1]
```diff
@@ -15,6 +15,7 @@
 --------------------------------------------------------------------------------
 
 -- stable release?
+local ignorelist = {} -- Ignore workaround for WG table.
```
Comment (sprunk) [inline LuaUI/cawidgets.lua]: Why? Just use WG

File: LuaUI/cawidgets.lua [hunk 2]
```diff
@@ -1370,6 +1378,16 @@ function widgetHandler:AddConsoleLine(msg, priority)
 			end
 		end
 	end
+    --Ignore's lobby blocker--
+    if newMsg.msgtype == 'autohost' and newMsg.argument and string.sub(newMsg.argument, 1, 1) == "<" then
```
Comment (sprunk) [inline LuaUI/cawidgets.lua]: Mixed tab and space indentation inside this block

File: LuaUI/Widgets/ignore.lua [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+local function ProcessString(str)
+	local strtbl = {}
+	for w in string.gmatch(str, "%S+") do
+		strtbl[#strtbl+1] = w
+	end
+	return strtbl
+end
+
+if VFS.FileExists("ignorelist.lua") then
+	WG.IgnoreList = VFS.Include("ignorelist.lua")
```
Comment (sprunk) [inline LuaUI/Widgets/ignore.lua]: Should use widget:GetConfigData and widget:SetConfigData

File: LuaUI/cawidgets.lua (hunk 1 above)
Comment (ArchShaman) [inline LuaUI/cawidgets.lua]: Originally I did use WG. Googlefrog wanted a function.

File: LuaUI/cawidgets.lua (hunk 2 above)
Comment (ArchShaman) [inline LuaUI/cawidgets.lua]: Ok I'll fix that. It's mostly Komodo edit's fault. Likes to make new lines double spaced. (Also some of it is copy pasted)

File: LuaUI/Widgets/ignore.lua (hunk 3 above)
Comment (ArchShaman) [inline LuaUI/Widgets/ignore.lua]: Oh so it's the same thing sort of? It takes a table in SetConfigData?

File: LuaUI/cawidgets.lua (hunk 2 above)
Comment (ArchShaman) [inline LuaUI/cawidgets.lua]: Fixed and also fixed another block while I was at it. Thanks.

File: LuaUI/Widgets/ignore.lua (hunk 3 above)
Comment (sprunk) [inline LuaUI/Widgets/ignore.lua]: The main benefit to using the above is that it goes into the same file together with the rest of widget config. So if someone (in particular, Steam) wants to backup/copy/delete/create/etc. the config it's always just one file.

File: LuaUI/cawidgets.lua (hunk 1 above)
Comment (sprunk) [inline LuaUI/cawidgets.lua]: Alright.

Current implementation still relies on WG if other widgets want to to get the ignore list though. Perhaps add a wh.GetIgnoreList and make the WG stuff local.

File: LuaUI/Widgets/ignore.lua (hunk 3 above)
Comment (ArchShaman) [inline LuaUI/Widgets/ignore.lua]: Okay. So from what I understand of looking at the source code, this is a callin of some sort and I just declare variables inside of it to save/load stuff? For instance:
`function widget:GetConfigData()
    WG.IgnoreList = IgnoreList
  end

function widget:SetConfigData
    if not IgnoreList then
         IgnoreList = {}
    else
        IgnoreList = IgnoreList
    end
  end`

Comment (sprunk) [inline LuaUI/Widgets/ignore.lua]: No, you return a table and receive it back as an argument.

```
function widget:SetConfigData()
    local dataToSave = {}
    for ignoree in WG.IgnoreList do
        dataToSave[ignoree] = true
    end
    return dataToSave
end

function widget:GetConfigData(dataToLoad)
    for ignoree in dataToLoad do
        WG.IgnoreList[ignoree] = true
    end
end
```

Comment (ArchShaman) [inline LuaUI/Widgets/ignore.lua]: Aha, thanks. Was just testing how to get this to work. Was wondering why it wasn't working.

File: LuaUI/cawidgets.lua (hunk 1 above)
Comment (ArchShaman) [inline LuaUI/cawidgets.lua]: Ok, will do.

**Comment by ArchShaman** [comment]
This doesn't seem to be done yet. There's some bug left, sorry.

**Comment by GoogleFrog** [comment]
All done?

**Comment by sprunk** [comment]
http://zero-k.info/Forum/Thread/22868?postID=163058#163058 claims it doesn't work.

**Comment by ArchShaman** [comment]
The code itself works, and I can verify it works. That post is an error between keyboard and chair issue.

**Comment by ArchShaman** [comment]
Note the post following it: http://zero-k.info/Forum/Thread/22868?postID=163153#163153
Note that the user in question also typed " /ignore _Shaman" (thus sending it as a say command instead of an actual text command). This leads to a conclusion of purposefully screwing up in order to mess with me. Proof:
> [4:20 PM]  Springiee [entropy3] /ignore _Shaman
Also note that this user has been nothing but overtly hostile towards me in the past. I would discredit this claim pretty quickly.

**Comment by sprunk** [comment]
Get somebody to explicitly confirm.

**Comment by ArchShaman** [comment]
This should be good now.

**Comment by sprangg** [comment]
Yeah there was a bug that made it not work for entropy3 or me, fixed above.

**Comment by fxrs** [comment]
This widget works without problems on my machine.

**Comment by Licho1** [comment]
In master, ignored users are auto set up in user params "ignored" in start script, and are easier accessible.

**Comment by ArchShaman** [comment]
Alright cool. Will remove parsing.

File: LuaUI/cawidgets.lua [hunk 4]
```diff
@@ -388,6 +388,11 @@ function widgetHandler:Initialize()
     isStable = true
   end
 
+  -- Add ignorelist --
+  local customkeys = select(10, Spring.GetPlayerInfo(Spring.GetMyPlayerID()))
+  if customkeys["ignored"] then
+    ignorelist = customkeys["ignored"]
```
Comment (sprunk) [inline LuaUI/cawidgets.lua]: Won't work. This value is not a table but a string ("name1,name2,name3"). You'll need to parse it to extract the values.

**Comment by ArchShaman** [comment]
Ah thanks for the heads up

**Comment by ArchShaman** [comment]
Currently hunting down the last bug: random numbers are added to the ignore list.

**Comment by GoogleFrog** [comment]
Why is this now about two unrelated things?

**Comment by ArchShaman** [comment]
Couldn't open another pr for it and at the time licho wanted it changed. Don't know how to undo it now.

**Comment by sprunk** [comment]
Git rebase

File: LuaUI/Widgets/gui_epicmenu.lua [hunk 5]
```diff
@@ -2346,7 +2346,7 @@ local function MakeQuitButtons()
 		icon = imgPath..'epicmenu/whiteflag_check.png',
 		OnChange = function()
 				if not (Spring.GetSpectatingState() or PlayingButNoTeammate() or isMission) then
-					spSendCommands("say !voteresign")
+					spSendCommands("say !poll resign")
```
Comment (Licho1) [inline LuaUI/Widgets/gui_epicmenu.lua]: you can use !resign directly  it will auto poll

**Comment by ArchShaman** [comment]
>Git rebase

What?

**Comment by GoogleFrog** [comment]
You should make a branch in your repository for each PR you make. You can then checkout each branch to make changes to it.

**Comment by ArchShaman** [comment]
>You should make a branch in your repository for each PR you make. You can then checkout each branch >to make changes to it.

Oh, thanks for the info! Okay, will remember that. Kinda new to github still. Should compile this information on the wiki under the development category.

**Comment by ArchShaman** [comment]
Closing this and remaking pr with new info in mind.

**Pull request closed without merging**